cmd/reference-addon-manager/main.go

```diff
@@ -92,14 +98,31 @@ func main() {
 		}
 	}
 
-	// if err = (&controllers.AddonReconciler{
-	// 	Client: mgr.GetClient(),
-	// 	Log:    ctrl.Log.WithName("controllers").WithName("Addon"),
-	// 	Scheme: mgr.GetScheme(),
-	// }).SetupWithManager(mgr); err != nil {
-	// 	setupLog.Error(err, "unable to create controller", "controller", "Addon")
-	// 	os.Exit(1)
-	// }
+	// the following sectiom hooks up a heartbeat reporter with the current addon/operator
+	///////////////////////////////////////////////////// -------------------------------- /////////////////////////////////////////////////////
+	addonName := "reference-addon"
+	// the following 'handleAddonInstanceConfigurationChanges' function can be absolutely anything depending how reference-addon would want to deal with AddonInstance's configuration change
+	handleAddonInstanceConfigurationChanges := func(addonsv1alpha1.AddonInstanceSpec) {
+		fmt.Println("Handling AddonInstance's configuration changes, whooossh!!!")
+	}
+
+	// setup the heartbeat reporter
+	heartbeatCommunicatorCh, err := utils.SetupHeartbeatReporter(mgr, addonName, handleAddonInstanceConfigurationChanges)
+	if err != nil {
+		setupLog.Error(err, "unable to setup heartbeat reporter")
+		os.Exit(1)
+	}
+	///////////////////////////////////////////////////// -------------------------------- /////////////////////////////////////////////////////
+
+	if err = (&controllers.ReferenceAddonReconciler{
+		Client:                       mgr.GetClient(),
+		Log:                          ctrl.Log.WithName("controllers").WithName("ReferenceAddon"),
+		Scheme:                       mgr.GetScheme(),
+		HeartbeatCommunicatorChannel: heartbeatCommunicatorCh, // linking the heartbeat communicator channel with the reconciler
```

Comment:
Ohh yeah this makes sense. The way I visualise a convenient way for heartbeat reporting is a central loop running concurrently which is reporting heartbeats periodically, and various sections of the addon's codebase are, from time to time, communicating to that central loop which heartbeat to communicate.

I tried to achieve that via channels and had thought about the concerns of channel congestions and blocks but tried to workaround them via timeout-backed channel-sends. But still the other concerns you mentioned are legit.

After thinking a bit about it, I feel that instead of having a channel (`heartbeatCommunicatorCh`) as the point of communication between various sections of the code and the central loop sending heartbeats, we can instead simply have a pointer to a shared variable which encapsulates which heartbeat to send. And the pointer to this shared variable would be an attribute of the ReferenceAddonReconciler struct. This means that across the codebase in various sections, the pointer to that shared variable can be accessed via the ReferenceAddonReconciler object `r`

If this is sounding a bit complicated then, let me come up with a super quick prototype to demostrate this :)

Reply:
+1 for prototype. :)
Not quite sure why you emphasize pointers so much in best case we have interfaces on the consumer side to decouple everything and make unit testing and refactoring easier.

Reply:
Ohh yes, scrape off that pointer thingy, I was thinking out loud at that moment, my bad :)

Reply:
Just a quick explanation in the approach I chose.
So, from now, the `ReferenceAddonReconciler` struct will have an attribute called `latestHeartbeat`. This is the so-called shared variable which was talking about.

The central loop (which would report heartbeat) will be defined under main.go and that loop, at every iteration, will access the `r.latestHeartbeat` (actually the getter method of it), and report that condition.

Whereas in various sections of code, whenever something happens and the Addon starts wanting to report a different heartbeat (say, an unhealthy heartbeat), all it has to do is to call `r.SetHeartbeat(...)`, that's it!
This method will update the `r.latestHeartbeat` value and from the next iteration of central loop, it will start reporting this new heartbeat, until changed again.